### ISSUE TITLE
twisted.web.http.Request.unregisterProducer has incomplete test coverage

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6121]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#6121|
|Type|defect|
|Created|2012-10-16 13:12:02Z|
|Branch|https://github.com/twisted/twisted/tree/request-unregisterproducer-coverage-6121|

This method is not tested at all by `test_http.py`.  Part of it is tested by something else in `twisted.web.test` though (not sure by what, though).  It should be completed tested by tests in `test_http.py`.  (found while porting http.py to Python 3)

Attachments:

* [6121.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/9de/9de33488cb4c45b2483aa4a6c42377535d1a1f62/d2c2e8224b6fd6a7de43e12faf320d25a6441cc6.patch) (2374 bytes) - added by moijes12 on 2013-04-04 12:59:12Z - 
* [6121.2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/9de/9de33488cb4c45b2483aa4a6c42377535d1a1f62/f0520276f02c86a752aeb6add05edc580bcb52ee.patch) (2132 bytes) - added by moijes12 on 2013-07-17 09:42:04Z - 
* [6121.3.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/9de/9de33488cb4c45b2483aa4a6c42377535d1a1f62/c6dccecc42b646dd2c8e70f4e2349dad36106032.patch) (2306 bytes) - added by moijes12 on 2013-07-20 16:59:36Z - Added 6121.misc file.

<details><summary>Searchable metadata</summary>

```
trac-id__6121 6121
type__defect defect
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_request_unregisterproducer_coverage_6121 branches/request-unregisterproducer-coverage-6121
branch_author__rwall rwall
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1350393122000000 1350393122000000
changetime__1378640277000000 1378640277000000
version__None None
owner__exarkun exarkun
cc__jknight
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="moijes12's avatar" src="https://avatars.githubusercontent.com/u/4566851?s=50" width="50" height="50">](https://github.com/moijes12)<a name="note_2"></a>|@moijes12 commented|
|-|-|

The attached patch contains tests for 4 scenarios :

 1. For a queued request with streaming set to false
 2. For a queued request with streaming set to true
 3. For a non-queued request with streaming set to false
 4. For a non-queued request with streaming set to true

The non-queued request tests use StringTransport (see [#6401](https://github.com/twisted/twisted/issues/6401)).

The set of regression tests have been run and no unexpected failures or errors are reported.

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_3"></a>|@wallrj set owner to @wallrj<br>@wallrj set status to `assigned`|
|-|-|

Reviewing...

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_4"></a>|@wallrj commented|
|-|-|

(In [38230]) Branching to 'request-unregisterproducer-coverage-6121'

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_5"></a>|@wallrj commented|
|-|-|

(In [38231]) Apply 6121.patch from moijes12. Refs [#6121](https://github.com/twisted/twisted/issues/6121)

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_6"></a>|@wallrj removed owner<br>@wallrj set status to `new`|
|-|-|

## Code Review

Thanks moijes12 for the new tests.

Notes:
 * I've applied your patch to a branch and run a build (there are some
   important failures - see below)
   * log:branches/request-unregisterproducer-coverage-6121
   * http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/request-unregisterproducer-coverage-6121
 * New tests pass for me.
 * Coverage of the unregisterProducer method does now seem complete. Here's what
   I did:
   ```
   $ coverage run --source twisted.web.http ./bin/trial twisted.web.test.test_http
   $ coverage html -d coverage-html
   $ firefox  coverage-html/twisted_web_http.html
   ```

Please answer or address the following comments:
 1. Patch didn't apply cleanly - it's probably ok - but please double check.
    ```
     patch -p0 < ~/Downloads/6121.patch
     patching file twisted/web/test/test_http.py
     Hunk #2 succeeded at 1699 with fuzz 1 (offset 46 lines).
    ```
 2. Patch contained trailing whitespace: https://buildbot.twistedmatrix.com/builders/twistedchecker/builds/619/steps/run-twistedchecker/logs/new%20twistedchecker%20errors
 3. Missing news file - I've added a blank .misc to the branch for you.
 4. DummyRequest is imported but unused: https://buildbot.twistedmatrix.com/builders/pyflakes/builds/330/steps/pyflakes/logs/new%20pyflakes%20errors
 5. Typo "requests'" should be "request's"
 6. There is a call to unregisterProducer in _cleanup which isn't
    tested. I'm not sure whether that should be dealt with in this
    branch. Perhaps another reviewer will comment on that.
 7. Maybe test that calling unregisterProducer twice (or without
    having first called registerProducer) does not raise an exception.
 8. I've struggled to trace through the http.Request code, but I notice you are assigning
    req.transport = StringTransport() in each of your *non-queued* tests,
    so StringTransport.(un)registerProducer should be called by Request.(un)registerProducer in those
    tests, but StringTransport.(un)registerProducer don't exist. So
    when I try and recreate this in a python shell I get
    "AttributeError: '_io.BytesIO' object has no attribute
    'registerProducer'". Maybe I'm just confused.

Anyway, thanks again. Please address or answer the numbered points above and
submit another patch (against the branch) for review. And someone more
familiar with twisted.web should probably do the next review.

-RichardW.

**Comment by twisted-trac** [comment]
|[<img alt="moijes12's avatar" src="https://avatars.githubusercontent.com/u/4566851?s=50" width="50" height="50">](https://github.com/moijes12)<a name="note_7"></a>|@moijes12 commented|
|-|-|

Hi

I have tried to address points 2,3,4 and 5. For 1, I'm not sure what needs to be done and as for 6,7 and 8, I'll have a look and get back later. I have attached a patch that I built against the trunk.

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_8"></a>|@habnabit commented|
|-|-|

The most recent attached patch does not add a news file.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set owner to @exarkun|
|-|-|

  1. For the non-queued case, `False` should be passed in for the `queued` parameter to `Request`, not `None`.
  1. A sort of nice property for test methods to have is for them to only make one assertion.  That way if they fail, you know everything that's wrong right away.  Otherwise, the first assertion might fail - then you fix it; then the second assertion might still fail.
  1. It would probably be a good idea to make some assertions about the transport's producer in the queued case.  Otherwise the conditional in `unregisterProducer` isn't really being tested (it is executed but no assertions depend on it being *correct*).

These are mostly minor points.  I'll address these in the branch and then look at rwall's later points.  Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

> There is a call to unregisterProducer in _cleanup which isn't tested. I'm not sure whether that should be dealt with in this branch. Perhaps another reviewer will comment on that.

I think it's alright to test that later on.  This ticket can focus on just the implementation of `unregisterProducer`.

> Maybe test that calling unregisterProducer twice (or without having first called registerProducer) does not raise an exception.

Well.. maybe.  It's sort of tempting.  On the other hand, `IConsumer` doesn't specify what happens when you try to unregister a producer when one isn't registered.  Some implementations silently accept it as a no-op, others don't (for example, `StringTransport` raises an exception :).

I feel like leaving this issue alone for now.

> I've struggled to trace through the http.Request code, but I notice you are assigning req.transport = StringTransport() in each of your *non-queued* tests, so StringTransport.(un)registerProducer should be called by Request.(un)registerProducer in those tests, but StringTransport.(un)registerProducer don't exist. So when I try and recreate this in a python shell I get "AttributeError: '_io.BytesIO' object has no attribute 'registerProducer'". Maybe I'm just confused.

Hm?  `StringTransport.unregisterProducer` does exist.  Maybe you picked up `StringTransport` from the wrong place?  (Notice it is from twisted.test.proto_helpers, not twisted.web.http)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun set status to `closed`|
|-|-|

(In [39881]) Merge request-unregisterproducer-coverage-6121

Author: moijes12, exarkun
Reviewer: rwall, exarkun
Fixes: [#6121](https://github.com/twisted/twisted/issues/6121)

Add some missing test coverage for `twisted.web.http.Request.unregisterProducer`.
